### ISSUE TITLE
`enum` forced unknown and organization

### ISSUE THREAD

**Issue by LePips** [CreateAPI/CreateAPI#130]
## Forced unknown
Since we are creating API interfaces, new cases may be created by available from the server that aren't available on a current version of a developer's SDK or even their available schema. Additionally, only [Apple can create nonfrozen enums](https://docs.swift.org/swift-book/ReferenceManual/Statements.html) by my understanding so developers cannot use `@unknown default`. So, if an app with an out-of-date SDK attempts to decode an unknown case for a property or return type that cannot be optional, it will crash.

To fix this issue, create a new config option `enumForcedUnknown` that adds a new case to every enum which represents an unknown case. The name is open for improvement, but I think that it should stand out from typical Swift formatting in a way that ensures it will be different from any case in the schema **and** tells the developer that this is strongly unknown.
```swift
enum MyTypes {
    case foo
    case bar
    case forced_unknown // added unknown case
}
```

## Organization

Smaller pitch for an organization option to put `enum`s in a separate folder: `Enums`. Just as an organizational measure this can help a lot to separate the `struct/class` entities from other Swift types. This would only apply to the split generation.

**Comment by LePips** [comment]
> The JSON Schema/OpenAPI specification does not support any differentiation between frozen or unfrozen enums and technically when you are using a different schema, it should be versioned differently.

No, that's because this lands in the territory of how a Swift developer uses this schema. This isn't about frozen/unfrozen enums _within_ the same version of the schema but across different schema versions. I still very much consider this "unfrozen" as an API is very much open to evolution. It's somewhat obvious that an application should not crash due to unexpected responses from an API, at least as much as it can.

---
Basic example, if `Pet.type` was of type (not optional) enum with cases `.dog, .cat` in V1 and then V1.1 added `.bird` to the available cases:

1 - if the developer didn't update the app's API to V1.1 it would crash for everyone
2 - if the developer did update the app's API but people haven't updated the app, it will crash for them

Instead the application just probably wouldn't show these unknown types and they can only be shown to people who update their app.

---

We don't necessarily have the responsibility to handle this, after all that is what the schema _explicitly_ states, but it would definitely help us developers _use_ an API. This kind of addition (among all the other options like exclusions and renamings I argue) is strictly Swift developer facing and does not actually go _against_ the schema, it just helps us work with safety and convenience.

> `RawRepresentable` instead of enum

I think this can be annoying because then one would have to switch against `String`, which would always require a `default` case. I was thinking something with a decoder that could fallback to the unknown case if that's included, hence why we should choose a name that will probably not conflict with existing cases but that can also be customizable.

**Comment by marcprux** [comment]
I just ran into this issue in a project that uses CreateAPI to access Apple's App Store Connect API: https://github.com/AvdLee/appstoreconnect-swift-sdk/pull/197#partial-pull-merging. 

I think the `RawRepresentable` String wrapper solution is better than the `forced_unknown` enum case, since client code will still be able to access the `rawValue` of the enumeration, and it lends itself to adding undocumented cases using extensions, like so:

```swift
struct BundleIdPlatform : Codable, RawRepresentable, Equatable { 
    var rawValue: String
    static let ios = Self(rawValue: "IOS")
    static let macos = Self(rawValue: "MAC_OS")
}

extension BundleIdPlatform {
    // unhandled cases can be manually added with an extension
    static let universal = Self(rawValue: "UNIVERSAL")
}
```

But I also think there are some cases where an enum might make more sense, such as when the API documents an explicit guarantee that the enumerated values will never change. So perhaps this could be added as a configuration attribute like `entities.openEnumerations` listing the type names that should be instead implemented as a `RawRepresentable` struct.

Either way, the current implementation is a ticking time bomb. I'm surprised that more users of APIs like this are not experiencing mysterious parse errors from the lack of forward-compatibility for enums that add an additional case.

**Comment by LePips** [comment]
I see value now in creating the extensions from a `RawRepresentable`. That sounds good 👍

**Comment by marcprux** [comment]
While I'm spitballing ideas, maybe there could be a `entities.openEnumerationsByDefault` property that defaults to `false` in order to continue generating the same code as before (and thus not break client code that might be relying on switch statements). When `entities.openEnumerationsByDefault` is `true`, it would behave as if all of the types were listed in the `entities.openEnumerations` property, and so generate the `RawRepresentable` implementations by default.

This would enable projects to easily choose between enabling open enumerations by default, or else progressively opt-in to using extensible enumerations with the `entities.openEnumerations`.

I suppose, for completeness, there should also be a `entities.frozenEnumerations` list that allows a user to override the `entities.openEnumerationsByDefault = true` behavior for a set of specific types. This could be useful when an API contains many open enumerations, but has a few that are documented to never change (e.g., some logically-frozen enum like `cardinalDirection = ["north", "south", "east", "west"]`).

**Comment by LePips** [comment]
With https://github.com/CreateAPI/CreateAPI/pull/111 we can now condense the options into two:
- `entities.openEnumerations`
- `entities.frozenEnumerations`

We would of course just include the checks that neither of these are true or false at the same time, they must independent binary values. Haven't thought through this yet but that sounds good as well

**Comment by liamnichols** [comment]
I don't think #111 will help in this instance actually because that only applies when using a Set of values that are `CaseIterable` (like `mutableProperties`).

We can however probably follow the same pattern that we do with `entities.defaultType` and `entities.typeOverrides`.

Something like `defaultEnumBehaviour` and `enumBehaviourOverrides`? 

### Disable

**Before**
```yaml
isGeneratingEnums: false
```

**After**
```yaml
enumOutput: string
```

### Closed/Frozen/Normal

**Before**
```yaml
isGeneratingEnums: true
```

**After**
```yaml
enumOutput: enum
```

### Open/RawRepresentable

```yaml
enumOutput: rawRepresentable
```

### Overrides

```yaml
enumOutput: enum
enumOutputOverrides:
  FooEnum: rawRepresentable
  BarEnum: string
```

I'm still not convinced about the naming that I've suggested above, but having the two options like this seems to make the most sense I think? It also covers all the bases described above. 

The one problem that I can think of though is how to reference an enum in `enumOutputOverrides` since enums aren't defined as schemas. I'm not completely sure how CreateAPI goes about nesting them/generating their names? Maybe it's derived from the property name and then nested within another entity? Does that always happen?

Do we want to try and make sure that we only reference types in the context of the spec? In that case we might need to use the `FooType.barProperty` syntax to reference the property that is the enum in the schema?

**Comment by liamnichols** [comment]
Oh I was wrong, you can define enums in `components.schemas` as part of the spec - https://swagger.io/docs/specification/data-models/enums/

**Comment by liamnichols** [comment]
It turns out that enums can be `integer` and other types as well so using `string` to represent `isGeneratingEnums: false` isn't quite right

**Comment by LePips** [comment]
Mirroring the type overrides sounds a lot neater.

**Comment by LePips** [comment]
Building off of the earlier comment, I think we can make a separate section for enum generation just like paths and entities. From a schema perspective they're a different type of object from an entity and from our perspective they don't match anything in the `entities` section.

```yaml
enums:
  defaultType: enum || static
  typeOverrides:
  myEnum: enum || static
  # I bet there are a few more we can make here like init options
  # or even turning off case iterable for enum generation and/or
  # providing the equivalent for static generation.
  # Also additional protocol conformances if they desire?
```

This would overall generalize the work for this.

**Comment by liamnichols** [comment]
Interesting. With that in mind, does it also mean that we should include enums in the new `generate` (#134) option instead of the `generateEnums` option? 

```yaml
generate: [entities, enums]
enums:
  defaultType: struct # (== RawRepresentable)
  typeOverrides:
    FrozenOption: enum
  # out of scope, but as an example:
  protocols:
  - CaseIterable
```

**Comment by liamnichols** [comment]
It feels like it would make sense to incorporate `enums` under `generate` regardless to if we bring in more options or not actually. I'll probably go ahead and do that in my change anyway unless there is a good reason not to?

**Comment by LePips** [comment]
Sensible to me

**Comment by liamnichols** [comment]
📝 Looks like this could have the potential to provide a workaround for AvdLee/appstoreconnect-swift-sdk#197 as well 👍

**Comment by nickasd** [comment]
This would also be very much welcome for App Store Connect API. Currently it contains several enums representing software platforms, screenshot types and so on that can potentially change whenever an Apple product with a new operating system or screen size is launched.
